### PR TITLE
Implement drop-muon optimizer

### PR DESCRIPTION
**Pull request by Ffinnis**
**Drop-Muon**

So, the general idea of the drop-muon optimizer is to make partial network updates based on a randomized schedule. We choose random subset of active layers and update parameters of only this layers, while others are frozen

Realization was generally described in the paper https://arxiv.org/pdf/2510.02239, in core it's just freeze some of the layers and update their params later

As the last variation I choose epoch-shift distribution to reorder priority based on the current stage of training: in the start of training we start from top-layers and then go to the bottom layers to train everything

**Benchmarks**

My baseline with RTX PRO 6000 (8M):
```
======================================================================
 SPEEDRUN RESULTS
======================================================================
Warmup & Setup:                  4s 79ms
Training Time (⏱️ Speedrun):      51s 253ms
Total Tokens:                    8,011,776
----------------------------------------------------------------------
Final Train Loss:                4.7344
Final Val Loss:                  4.8384
Final Val Accuracy:              0.2453
======================================================================
```

Drop-Muon (8M):
```
======================================================================
 SPEEDRUN RESULTS
======================================================================
Warmup & Setup:                  4s 48ms
Training Time (⏱️ Speedrun):      45s 580ms
Total Tokens:                    8,011,776
----------------------------------------------------------------------
Final Train Loss:                4.8788
Final Val Loss:                  4.9812
Final Val Accuracy:              0.2339
======================================================================
```

For 8M it not a great deal, we lost 0.15 loss (3%) and gain 5s (12%)

Same repeats for 100M tokens

Base:
```
======================================================================
 SPEEDRUN RESULTS
======================================================================
Warmup & Setup:                  4s 14ms
Training Time (⏱️ Speedrun):      9m 35s 480ms
Total Tokens:                    100,007,936
----------------------------------------------------------------------
Final Train Loss:                3.6732
Final Val Loss:                  3.7172
Final Val Accuracy:              0.3501
======================================================================
```

Drop-Muon:
```
======================================================================
 SPEEDRUN RESULTS
======================================================================
Warmup & Setup:                  4s 41ms
Training Time (⏱️ Speedrun):      8m 22s 179ms
Total Tokens:                    100,007,936
----------------------------------------------------------------------
Final Train Loss:                3.7262
Final Val Loss:                  3.7857
Final Val Accuracy:              0.3440
======================================================================
```

It's 1m 13s difference (~12%) but we again lost some val_loss 0.7 (2%), I think it's a cool trade. Probably with muon-drop optimizer we can run ~115M vs 100M tokens with the same time for both runs

### CODE REVIEW

**Comment by vukrosic** [comment]
@Ffinnis See that it gives a bit worse loss even though it's faster, not sure if this will prove bad on bigger dataset

What do you think about this? Maybe I should train on 100M and see?

image

**Comment by Roma-Potapov** [comment]
> @Ffinnis See that it gives a bit worse loss even though it's faster, not sure if this will prove bad on bigger dataset
> 
> 
> 
> What do you think about this? Maybe I should train on 100M and see?
> 
> 
> 
> image
> 
> 

Yeah, try on 100M tokens, the gap in loss is ok for this optimizer, because we train less layers at the same time, but the speed difference is big

I got lower gap between drop muon and muon on 100M than on 8M

**Comment by Ffinnis** [comment]
> @Ffinnis See that it gives a bit worse loss even though it's faster, not sure if this will prove bad on bigger dataset
> 
> What do you think about this? Maybe I should train on 100M and see?
> 
> image

@vukrosic So, what's the conclusion? Will we merge this changes into main branch or save for further investigation? I wrote my opinion from this account @Roma-Potapov previously

**Comment by vukrosic** [comment]
It seems like it's not worthed, if we look at val loss vs training time, baseline is better, it learns faster.

image

image
